pageserver/src/walredo.rs

```diff
@@ -165,16 +191,29 @@ impl WalRedoManager for PostgresRedoManager {
             records,
         };
 
-        // launch the WAL redo process on first use
-        let mut process_guard = self.process.lock().unwrap();
-        if process_guard.is_none() {
-            let p = self.runtime
-                .block_on(PostgresRedoProcess::launch(self.conf, &self.tenantid))?;
-            *process_guard = Some(p);
-        }
-        let process = (*process_guard).as_ref().unwrap();
+        start_time = Instant::now();
+        let result = {
+            let mut process_guard = self.process.lock().unwrap();
+            lock_time = Instant::now();
+
+            // launch the WAL redo process on first use
+            if process_guard.is_none() {
+                let p = self
+                    .runtime
+                    .block_on(PostgresRedoProcess::launch(self.conf, &self.tenantid))?;
+                *process_guard = Some(p);
+            }
+            let process = (*process_guard).as_ref().unwrap();
+
+            self.runtime
+                .block_on(self.handle_apply_request(&process, &request))
+        };
+        end_time = Instant::now();
```

Comment:
For this we have a very convenient method [observe_closure_duration](https://docs.rs/prometheus/0.12.0/prometheus/struct.Histogram.html#method.observe_closure_duration).

```rust
let result = WAL_REDO_TIME.observe_closure_duration(|| {
    // notice how the closure result is getting passed through
    let mut process_guard =
        WAL_REDO_WAIT_TIME.observe_closure_duration(|| self.process.lock().unwrap());

    // launch the WAL redo process on first use
    if process_guard.is_none() {
        let p = self
            .runtime
            .block_on(PostgresRedoProcess::launch(self.conf, &self.tenantid))?;
        *process_guard = Some(p);
    }
    let process = process_guard.as_ref().unwrap();

    self.runtime
        .block_on(self.handle_apply_request(&process, &request))
});
```

Reply:
I wanted to use the same timestamp as the end of the "waiting" phase and as the start of the "redo" phase. And I wanted to store the observation only after releasing the lock, to keep the locked portion as short as possible. Probably doesn't matter much in practice.